From 0303f4a74a968a15d79d9a50c4d8fe3d1e71a3dd Mon Sep 17 00:00:00 2001
From: Ruchi Kandoi <kandoiruchi@google.com>
Date: Thu, 24 Jan 2019 13:39:18 -0800
Subject: [PATCH 3/5] Prevent OOB error in rw_i93_sm_detect_ndef()

Bug: 121260197
Test: NFC tag reading
Change-Id: I9168e338a802c43122b252e895fa4dffcd7080f4
Merged-In: I9168e338a802c43122b252e895fa4dffcd7080f4
(cherry picked from commit 9939edeb9fd6b118c0594c0d07459d4042ed0017)
---
 src/nfc/tags/rw_i93.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/src/nfc/tags/rw_i93.c b/src/nfc/tags/rw_i93.c
index 68d61cf..605ce1b 100644
--- a/src/nfc/tags/rw_i93.c
+++ b/src/nfc/tags/rw_i93.c
@@ -23,6 +23,7 @@
  *  mode.
  *
  ******************************************************************************/
+#include <log/log.h>
 #include <string.h>
 #include "nfc_target.h"
 #include "bt_types.h"
@@ -1536,6 +1537,12 @@ void rw_i93_sm_detect_ndef (BT_HDR *p_resp)
     RW_TRACE_DEBUG1 ("rw_i93_sm_detect_ndef () sub_state:0x%x", p_i93->sub_state);
 #endif
 
+    if (length == 0)
+    {
+        android_errorWriteLog(0x534e4554, "121260197");
+        rw_i93_handle_error(NFC_STATUS_FAILED);
+        return;
+    }
     STREAM_TO_UINT8 (flags, p);
     length--;
 
@@ -1558,7 +1565,12 @@ void rw_i93_sm_detect_ndef (BT_HDR *p_resp)
     switch (p_i93->sub_state)
     {
     case RW_I93_SUBSTATE_WAIT_UID:
-
+        if (length < (I93_UID_BYTE_LEN + 1))
+        {
+            android_errorWriteLog(0x534e4554, "121260197");
+            rw_i93_handle_error(NFC_STATUS_FAILED);
+            return;
+        }
         STREAM_TO_UINT8 (u8, p); /* DSFID */
         p_uid = p_i93->uid;
         STREAM_TO_ARRAY8 (p_uid, p);
-- 
2.7.4

